fabric-registry-sync-v0/src/client/java/net/fabricmc/fabric/impl/client/registry/sync/ClientRegistrySyncHandler.java

```diff
@@ -0,0 +1,198 @@
+import org.jetbrains.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.minecraft.registry.Registries;
+import net.minecraft.registry.Registry;
+import net.minecraft.screen.ScreenTexts;
+import net.minecraft.text.MutableText;
+import net.minecraft.text.Text;
+import net.minecraft.util.Formatting;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.thread.ThreadExecutor;
+
+import net.fabricmc.fabric.api.event.registry.RegistryAttribute;
+import net.fabricmc.fabric.impl.registry.sync.RegistrySyncManager;
+import net.fabricmc.fabric.impl.registry.sync.RemapException;
+import net.fabricmc.fabric.impl.registry.sync.RemappableRegistry;
+import net.fabricmc.fabric.impl.registry.sync.packet.RegistryPacketHandler;
+
+public final class ClientRegistrySyncHandler {
+	private static final Logger LOGGER = LoggerFactory.getLogger(ClientRegistrySyncHandler.class);
+
+	private ClientRegistrySyncHandler() {
+	}
+
+	public static <T extends RegistryPacketHandler.RegistrySyncPayload> CompletableFuture<Boolean> receivePacket(ThreadExecutor<?> executor, RegistryPacketHandler<T> handler, T payload, boolean accept) {
+		handler.receivePayload(payload);
+
+		if (!handler.isPacketFinished()) {
+			return CompletableFuture.completedFuture(false);
+		}
+
+		if (RegistrySyncManager.DEBUG) {
+			String handlerName = handler.getClass().getSimpleName();
+			LOGGER.info("{} total packet: {}", handlerName, handler.getTotalPacketReceived());
+			LOGGER.info("{} raw size: {}", handlerName, handler.getRawBufSize());
+			LOGGER.info("{} deflated size: {}", handlerName, handler.getDeflatedBufSize());
+		}
+
+		RegistryPacketHandler.SyncedPacketData data = handler.getSyncedPacketData();
+
+		if (!accept) {
+			return CompletableFuture.completedFuture(true);
+		}
+
+		return executor.submit(() -> {
+			if (data == null) {
+				throw new CompletionException(new RemapException("Received null map in sync packet!"));
+			}
+
+			try {
+				apply(data);
+				return true;
+			} catch (RemapException e) {
+				throw new CompletionException(e);
+			}
+		});
+	}
+
+	public static void apply(RegistryPacketHandler.SyncedPacketData data) throws RemapException {
+		// First check that all of the data provided is valid before making any changes
+		checkRemoteRemap(data);
+
+		for (Map.Entry<Identifier, Object2IntMap<Identifier>> entry : data.idMap().entrySet()) {
+			final Identifier registryId = entry.getKey();
+
+			Registry<?> registry = Registries.REGISTRIES.get(registryId);
+
+			// Registry was not found on the client, is it optional?
+			// If so we can just ignore it.
+			// Otherwise we throw an exception and disconnect.
+			if (registry == null) {
+				if (isRegistryOptional(registryId, data)) {
+					LOGGER.info("Received registry data for unknown optional registry: {}", registryId);
+					continue;
+				} else {
+					throw new RemapException("Received registry data for unknown registry: " + registryId);
+				}
+			}
```

Comment:
I believe this is duplicative, as `checkRemoteRemap` should've already thrown in this case?